docs/decisions/0040-filters-exception-handling.md

```diff
@@ -0,0 +1,199 @@
+}
+```
+
+Advantages:
+
+- Requires minimum changes to existing implementation and also it won't break existing filter users.
+- Similar to `IActionFilter` API in ASP.NET.
+- Scalable, because it will be possible to extend similar Context models for other type of filters when needed (prompt or function calling filters).
+
+Disadvantages:
+
+- Not .NET-friendly way of exception handling with `context.Exception = null` or `context.Exception = new AnotherException()`, instead of using native `try/catch` approach.
+
+### [Option 4] Change `IFunctionFilter` signature by adding `next` delegate.
+
+This approach changes the way how filters work at the moment. Instead of having two `Invoking` and `Invoked` methods in filter, there will be only one method that will be invoked during function execution with `next` delegate, which will be responsible to call next registered filter in pipeline or function itself, in case there are no remaining filters.
+
+Abstraction:
+
+```csharp
+public interface IFunctionFilter
+{
+    Task OnFunctionInvocationAsync(FunctionInvocationContext context, FunctionInvocationCallback next);
+}
+```
+
+Usage:
+
+```csharp
+public class MyFilter : IFunctionFilter
+{
+    public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, FunctionInvocationCallback next)
+    {
+        // Perform some actions before function invocation
+        await next(context);
+        // Perform some actions after function invocation
+    }
+}
+```
+
+Exception handling with native `try/catch` approach:
+
+```csharp
+public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, FunctionInvocationCallback next)
+{
+    try
+    {
+        await next(context);
+    }
+    catch (Exception exception)
+    {
+        this._logger.LogError(exception, "Something went wrong during function invocation");
+
+        // Example: override function result value
+        context.SetResultValue("Friendly message instead of exception");
+
+        // Example: Rethrow another type of exception if needed
+        throw new InvalidOperationException("New exception");
+    }
+}
+```
+
+Advantages:
+
+- Native way how to handle and rethrow exceptions.
+- Similar to `IAsyncActionFilter` and `IEndpointFilter` API in ASP.NET.
+- One filter method to implement instead of two (`Invoking/Invoked`).
+- No need in cancellation logic (e.g. `context.Cancel = true`). To cancel the operation, simply don't call `await next(context)`.
+
+Disadvantages:
+
+- Remember to call `await next(context)` manually in all filters. If it's not called, next filter in pipeline and/or function itself won't be called.
+- Since it's callback mechanism, for streaming scenarios only enumerator can be wrapped with filter. Which means that pre-invocation code will be applied (e.g. update arguments) but post-invocation code won't be applied (e.g. exception handling). `IEndpointFilter` in ASP.NET Minimal API doesn't handle exceptions in streaming scenario as well.
```

Comment:
> Is this a deal breaker for option 4?

It depends. Regarding calling `next` delegate manually - it's just a specifics of API, once you learn how to use it, it should not be a problem. 
Regarding post-invocation actions for streaming scenarios - this might be a problem in case we would like to support it. It's not supported in current version of filters. I implemented this functionality using Option 3, while it was hard to achieve same results with Option 4 because of the callback. So, I guess it depends on filter requirements for streaming scenarios.

Reply:
Resolving this comment, since with recent changes in Option 4 streaming scenarios for result overload and exception handling are supported as well.